benchmark/utils/generator.hpp

```diff
@@ -0,0 +1,278 @@
+struct DefaultSystemGenerator {
+    using index_type = IndexType;
+    using value_type = ValueType;
+    using Vec = vec<ValueType>;
+
+    static gko::matrix_data<ValueType, IndexType> generate_matrix_data(
+        rapidjson::Value& config)
+    {
+        if (config.HasMember("filename")) {
+            std::ifstream in(config["filename"].GetString());
+            return gko::read_generic_raw<ValueType, IndexType>(in);
+        } else if (config.HasMember("stencil")) {
+            return generate_stencil<ValueType, IndexType>(
+                config["stencil"].GetString(), config["size"].GetInt64());
+        } else {
+            throw std::runtime_error(
+                "No known way to generate matrix data found.");
+        }
+    }
+
+    static std::shared_ptr<gko::LinOp> generate_matrix(
+        std::shared_ptr<const gko::Executor> exec, rapidjson::Value& config)
+    {
+        auto data = generate_matrix_data(config);
+        return generate_matrix_with_format(
+            std::move(exec), config["optimal"]["spmv"].GetString(), data);
+    }
+
+    static std::shared_ptr<gko::LinOp> generate_matrix_with_format(
+        std::shared_ptr<const gko::Executor> exec,
+        const std::string& format_name,
+        const gko::matrix_data<ValueType, itype>& data)
+    {
+        return gko::share(
+            ::formats::matrix_factory(format_name, std::move(exec), data));
+    }
+
+    static std::shared_ptr<gko::LinOp> generate_matrix_with_default_format(
+        std::shared_ptr<const gko::Executor> exec,
+        const gko::matrix_data<ValueType, itype>& data)
+    {
+        return generate_matrix_with_format(std::move(exec), "coo", data);
+    }
+
+    static std::unique_ptr<Vec> create_multi_vector(
+        std::shared_ptr<const gko::Executor> exec, gko::dim<2> size,
+        ValueType value)
+    {
+        auto res = Vec::create(exec);
+        res->read(gko::matrix_data<ValueType, itype>(size, value));
+        return res;
+    }
+
+    static std::unique_ptr<Vec> create_multi_vector(
+        std::shared_ptr<const gko::Executor> exec, gko::dim<2> size,
+        gko::size_type stride)
+    {
+        auto res = Vec::create(exec, size, stride);
+        return res;
+    }
+
+    // creates a random multi_vector
+    static std::unique_ptr<Vec> create_multi_vector_random(
+        std::shared_ptr<const gko::Executor> exec, gko::dim<2> size)
+    {
+        auto res = Vec::create(exec);
+        res->read(gko::matrix_data<ValueType, itype>(
+            size,
+            std::uniform_real_distribution<gko::remove_complex<ValueType>>(-1.0,
+                                                                           1.0),
+            get_engine()));
+        return res;
+    }
+
+    // creates a zero vector
+    static std::unique_ptr<Vec> create_vector(
+        std::shared_ptr<const gko::Executor> exec, gko::size_type size)
+    {
+        return create_multi_vector(std::move(exec), gko::dim<2>{size, 1}, 1);
```

Comment:
the vector is most likely uninitialized, not zero!

Reply:
this is not even used...